compiler/p/codegen/OMRLoadStoreHandler.hpp

```diff
@@ -0,0 +1,319 @@
+#include "codegen/Register.hpp"
+#include "il/Node.hpp"
+#include "infra/Annotations.hpp"
+
+namespace OMR
+{
+
+namespace Power
+{
+
+/**
+ * \brief A memory reference alongside information about operations to perform when its lifetime
+ *        is complete.
+ */
+class NodeMemoryReference
+    {
+    TR::MemoryReference *_memRef;
+
+    public:
+    NodeMemoryReference() : _memRef(NULL) {}
+    explicit NodeMemoryReference(TR::MemoryReference *memRef) : _memRef(memRef) {}
+
+    TR::MemoryReference *getMemoryReference() { return _memRef; }
+    void decReferenceCounts(TR::CodeGenerator *cg) { return _memRef->decNodeReferenceCounts(cg); }
+    };
+
+/**
+ * \brief An extensible class containing static methods for evaluating loads and stores.
+ *
+ * This class contains static methods for assisting evaluators in generating load and store sequences. These methods
+ * are meant to be used from evaluators and are capable of some limited customization of how a load or store is
+ * performed to allow for codegen optimizations.
+ *
+ * This class is meant to be extended by downstream projects that require special handling for emitting load and store
+ * sequences for certain types of symbol references, e.g. unresolved symbols in OpenJ9.
+ *
+ * \warning Unless otherwise mentioned, methods on this class are free to emit arbitrary instructions at the append
+ *          cursor, and so should never be called during internal control flow.
+ */
+class OMR_EXTENSIBLE LoadStoreHandler
+    {
+    public:
+
+    /**
+     * \brief Generate a sequence of instructions to compute the effective address of a load, store, or loadaddr.
+     *
+     * This method generates a sequence of instructions at the append cursor that compute the effective address of a
+     * load, store, or loadaddr node. This is useful for evaluators that need to check or change the address of a
+     * load/store before it is performed, such as some forms of GC write barrier.
+     *
+     * For load nodes, the address computed by this method can be used with generateLoadAddressSequence to actually
+     * perform the load. For stores, the address computed by this method can be used with generateStoreAddressSequence
+     * to actually perform the store.
+     *
+     * \warning While it is possible to emit a load or store directly using the computed address, this should be done
+     *          with extreme caution and only in the most derived project. Downstream projects are free to customize
+     *          how load/store sequences are emitted, so this should never be done in projects where overriding
+     *          implementations could exist.
+     *
+     * \param cg The code generator
+     * \param addrReg The register into which the effective address should be computed
+     * \param node The load, store, or loadaddr node whose effective address should be computed
+     * \param extraOffset An extra offset in bytes to add to the effective address
+     */
```

Comment:
i haven't read other codes, but, based on this description, it looks like possible to regress performance, given you need to compute the whole target address before hand, losing the opportunity to take advantages of imm offset or index register form, at least for the special cases this will be used.

Reply:
This API is designed specifically for use in locations where it is necessary to know the address of the load/store before performing the load/store itself, i.e. cardcheck write barriers. In these cases, the whole target address was already being computed anyway, since it was needed by definition, so using this new API shouldn't regress performance.